File: back-bardak/src/test/java/kz/bardak/game/rules/DealEngineTest.java
package kz.bardak.game.rules;

import static kz.bardak.game.rules.DealStateFixture.aDeal;
import static org.assertj.core.api.Assertions.assertThat;

import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * Переходы автомата раздачи — §1.4, §2.1, §2.2, §4.2 правил.
 */
class DealEngineTest {

    private static final Card SEVEN_DIAMONDS = PipCard.of(Rank.SEVEN, Suit.DIAMONDS);
    private static final Card SEVEN_CLUBS = PipCard.of(Rank.SEVEN, Suit.CLUBS);
    private static final Card SEVEN_SPADES = PipCard.of(Rank.SEVEN, Suit.SPADES);
    private static final Card NINE_DIAMONDS = PipCard.of(Rank.NINE, Suit.DIAMONDS);
    private static final Card ACE_CLUBS = PipCard.of(Rank.ACE, Suit.CLUBS);
    private static final Card KING_CLUBS = PipCard.of(Rank.KING, Suit.CLUBS);

    private final DealEngine engine = DealEngine.withDefaults();

    @DisplayName("Should move the card from hand to the table When an attack is applied")
    @Test
    void shouldMoveTheCardFromHandToTheTableWhenAnAttackIsApplied() {
        final DealState state = aDeal().withHand(0, SEVEN_DIAMONDS, ACE_CLUBS).withHand(1, KING_CLUBS).build();

        final MoveResult result = engine.apply(state, new DealCommand.Attack(0, SEVEN_DIAMONDS));

        final DealState next = applied(result);
        assertThat(next.playerAt(0).hand()).containsExactly(ACE_CLUBS);
        assertThat(next.table()).singleElement().satisfies(slot -> {
            assertThat(slot.attack()).isEqualTo(SEVEN_DIAMONDS);
            assertThat(slot.isBeaten()).isFalse();
        });
        assertThat(next.phase()).isEqualTo(DealPhase.DEFEND);
        assertThat(events(result)).containsExactly(new DealEvent.CardAttacked(0, SEVEN_DIAMONDS));
    }

    @DisplayName("Should leave the state untouched When a command is rejected")
    @Test
    void shouldLeaveTheStateUntouchedWhenACommandIsRejected() {
        final DealState state = aDeal().withHand(2, SEVEN_DIAMONDS).withHand(1, KING_CLUBS).build();

        final MoveResult result = engine.apply(state, new DealCommand.Attack(2, SEVEN_DIAMONDS));

        assertThat(result).isEqualTo(MoveResult.rejected(RejectionReason.NOT_YOUR_TURN));
    }

    @DisplayName("Should mark the named slot as beaten When a defence is applied")
    @Test
    void shouldMarkTheNamedSlotAsBeatenWhenADefenceIsApplied() {
        final DealState state = aDeal()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(1, NINE_DIAMONDS)
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.Defend(1, NINE_DIAMONDS, SEVEN_DIAMONDS));

        final DealState next = applied(result);
        assertThat(next.table()).singleElement().satisfies(slot ->
                assertThat(slot.defenceCard()).contains(NINE_DIAMONDS));
        assertThat(next.anyCardBeatenThisRound()).isTrue();
        assertThat(next.phase()).isEqualTo(DealPhase.ATTACK);
        assertThat(next.playerAt(1).hand()).isEmpty();
    }

    @DisplayName("Should stay in defence When unbeaten cards remain on the table")
    @Test
    void shouldStayInDefenceWhenUnbeatenCardsRemainOnTheTable() {
        final DealState state = aDeal()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS, SEVEN_CLUBS)
                .withHand(1, NINE_DIAMONDS, PipCard.of(Rank.NINE, Suit.CLUBS))
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.Defend(1, NINE_DIAMONDS, SEVEN_DIAMONDS));

        assertThat(applied(result).phase()).isEqualTo(DealPhase.DEFEND);
    }

    @DisplayName("Should make the transferring player the attacker When an attack is transferred")
    @Test
    void shouldMakeTheTransferringPlayerTheAttackerWhenAnAttackIsTransferred() {
        final DealState state = aDeal()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(1, SEVEN_CLUBS)
                .withHand(2, ACE_CLUBS, KING_CLUBS)
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.Transfer(1, SEVEN_CLUBS));

        final DealState next = applied(result);
        assertThat(next.defenderSeat()).isEqualTo(2);
        assertThat(next.roundStarterSeat()).isEqualTo(1);
        assertThat(next.attackRightSeat()).isEqualTo(1);
        assertThat(next.attackCardCount()).isEqualTo(2);
        assertThat(events(result)).containsExactly(new DealEvent.AttackTransferred(1, 2, SEVEN_CLUBS));
    }

    @DisplayName("Should hand the attack right to the second neighbour When the round starter passes")
    @Test
    void shouldHandTheAttackRightToTheSecondNeighbourWhenTheRoundStarterPasses() {
        final DealState state = aDeal()
                .withSeats(4)
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(1, NINE_DIAMONDS)
                .withHand(2, SEVEN_CLUBS)
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.Pass(0));

        final DealState next = applied(result);
        assertThat(next.attackRightSeat()).isEqualTo(2);
        assertThat(next.hasPassed(0)).isTrue();
        assertThat(events(result)).containsExactly(new DealEvent.Passed(0), new DealEvent.AttackRightMoved(2));
    }

    @DisplayName("Should never give the attack right back When the passed player tries to attack again")
    @Test
    void shouldNeverGiveTheAttackRightBackWhenThePassedPlayerTriesToAttackAgain() {
        final DealState state = aDeal()
                .withSeats(4)
                .withAttackCards(SEVEN_DIAMONDS)
                .withPassed(0)
                .withHand(0, SEVEN_CLUBS)
                .withHand(1, NINE_DIAMONDS, ACE_CLUBS)
                .build();

        assertThat(engine.apply(state, new DealCommand.Attack(0, SEVEN_CLUBS)))
                .isEqualTo(MoveResult.rejected(RejectionReason.NOT_YOUR_TURN));
    }

    @DisplayName("Should close the round as beaten When the last attacker passes and nothing is unbeaten")
    @Test
    void shouldCloseTheRoundAsBeatenWhenTheLastAttackerPassesAndNothingIsUnbeaten() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withBeatenPair(SEVEN_DIAMONDS, NINE_DIAMONDS)
                .withHand(0, ACE_CLUBS)
                .withHand(1, KING_CLUBS)
                .withHand(2, SEVEN_CLUBS)
                .build();

        final DealState afterFirstPass = applied(engine.apply(state, new DealCommand.Pass(0)));
        assertThat(afterFirstPass.attackRightSeat()).isEqualTo(2);
        assertThat(afterFirstPass.table()).hasSize(1);

        final MoveResult result = engine.apply(afterFirstPass, new DealCommand.Pass(2));

        final DealState next = applied(result);
        assertThat(next.table()).isEmpty();
        assertThat(next.anyPileDiscarded()).isTrue();
        assertThat(next.roundStarterSeat()).isEqualTo(1);
        assertThat(next.defenderSeat()).isEqualTo(2);
        assertThat(next.phase()).isEqualTo(DealPhase.ATTACK);
        assertThat(events(result)).contains(new DealEvent.RoundBeaten(1, List.of(SEVEN_DIAMONDS, NINE_DIAMONDS)));
    }

    @DisplayName("Should wait for the defender When every attacker passed but cards are unbeaten")
    @Test
    void shouldWaitForTheDefenderWhenEveryAttackerPassedButCardsAreUnbeaten() {
        final DealState state = aDeal()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withPassed(2)
                .withHand(1, ACE_CLUBS)
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.Pass(0));

        final DealState next = applied(result);
        assertThat(next.phase()).isEqualTo(DealPhase.DEFEND);
        assertThat(next.table()).hasSize(1);
    }

    @DisplayName("Should keep the round alive for follow-up cards When the defender announces a take")
    @Test
    void shouldKeepTheRoundAliveForFollowUpCardsWhenTheDefenderAnnouncesATake() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(0, SEVEN_CLUBS)
                .withHand(1, KING_CLUBS)
                .withHand(2, SEVEN_SPADES)
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.Take(1));

        final DealState next = applied(result);
        assertThat(next.phase()).isEqualTo(DealPhase.TAKING);
        assertThat(next.table()).hasSize(1);
        assertThat(next.playerAt(1).hand()).containsExactly(KING_CLUBS);
        assertThat(events(result)).containsExactly(new DealEvent.TakeAnnounced(1));
    }

    @DisplayName("Should move the whole table into the hand When every attacker passed after the take")
    @Test
    void shouldMoveTheWholeTableIntoTheHandWhenEveryAttackerPassedAfterTheTake() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withBeatenPair(SEVEN_DIAMONDS, NINE_DIAMONDS)
                .withAttackCards(SEVEN_CLUBS)
                .withHand(0, ACE_CLUBS)
                .withHand(1, KING_CLUBS)
                .withHand(2, SEVEN_SPADES)
                .build();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));
        final DealState afterFirstPass = applied(engine.apply(announced, new DealCommand.Pass(0)));
        final MoveResult result = engine.apply(afterFirstPass, new DealCommand.Pass(2));

        final DealState next = applied(result);
        assertThat(next.playerAt(1).hand())
                .containsExactly(KING_CLUBS, SEVEN_DIAMONDS, NINE_DIAMONDS, SEVEN_CLUBS);
        assertThat(next.table()).isEmpty();
        assertThat(next.anyPileDiscarded()).isFalse();
    }

    @DisplayName("Should let attackers keep adding cards When the take has been announced")
    @Test
    void shouldLetAttackersKeepAddingCardsWhenTheTakeHasBeenAnnounced() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(0, SEVEN_CLUBS)
                .withHand(1, KING_CLUBS)
                .withHand(2, SEVEN_SPADES)
                .build();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));
        final DealState next = applied(engine.apply(announced, new DealCommand.Attack(0, SEVEN_CLUBS)));

        assertThat(next.phase()).isEqualTo(DealPhase.TAKING);
        assertThat(next.attackCardCount()).isEqualTo(2);
    }

    @DisplayName("Should ignore the taker's hand size When cards are added after the take")
    @Test
    void shouldIgnoreTheTakersHandSizeWhenCardsAreAddedAfterTheTake() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(0, SEVEN_CLUBS)
                .withHand(1)
                .withHand(2, SEVEN_SPADES)
                .build();

        assertThat(state.defender().defendableCards(true)).isZero();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));

        assertThat(engine.apply(announced, new DealCommand.Attack(0, SEVEN_CLUBS)).isApplied()).isTrue();
    }

    @DisplayName("Should still cap the follow-up cards at the round limit When the take has been announced")
    @Test
    void shouldStillCapTheFollowUpCardsAtTheRoundLimitWhenTheTakeHasBeenAnnounced() {
        final RulesConfig config = RulesConfig.defaults();
        final DealStateFixture fixture = aDeal().withEmptyDeck().withPhase(DealPhase.DEFEND);
        for (int index = 0; index < config.maxAttackFirstRound(); index++) {
            fixture.withAttackCards(PipCard.of(Rank.values()[index], Suit.DIAMONDS));
        }
        final DealState state = fixture
                .withHand(0, PipCard.of(Rank.SIX, Suit.HEARTS))
                .withHand(1, KING_CLUBS)
                .build();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));

        assertThat(engine.apply(announced, new DealCommand.Attack(0, PipCard.of(Rank.SIX, Suit.HEARTS))))
                .isEqualTo(MoveResult.rejected(RejectionReason.ATTACK_LIMIT_REACHED));
    }

    @DisplayName("Should refuse a defence When the take has already been announced")
    @Test
    void shouldRefuseADefenceWhenTheTakeHasAlreadyBeenAnnounced() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(0, SEVEN_CLUBS)
                .withHand(1, NINE_DIAMONDS)
                .build();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));

        assertThat(engine.apply(announced, new DealCommand.Defend(1, NINE_DIAMONDS, SEVEN_DIAMONDS)))
                .isEqualTo(MoveResult.rejected(RejectionReason.DEFENDER_ALREADY_TOOK));
    }

    @DisplayName("Should refuse a transfer When the take has already been announced")
    @Test
    void shouldRefuseATransferWhenTheTakeHasAlreadyBeenAnnounced() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(0, ACE_CLUBS)
                .withHand(1, SEVEN_CLUBS)
                .withHand(2, KING_CLUBS, SEVEN_SPADES)
                .build();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));

        assertThat(engine.apply(announced, new DealCommand.Transfer(1, SEVEN_CLUBS)))
                .isEqualTo(MoveResult.rejected(RejectionReason.DEFENDER_ALREADY_TOOK));
    }

    @DisplayName("Should skip the defender in the next round When the defender took the table")
    @Test
    void shouldSkipTheDefenderInTheNextRoundWhenTheDefenderTookTheTable() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withPhase(DealPhase.DEFEND)
                .withAttackCards(SEVEN_DIAMONDS)
                .withHand(0, ACE_CLUBS)
                .withHand(1, KING_CLUBS)
                .withHand(2, SEVEN_SPADES)
                .build();

        final DealState announced = applied(engine.apply(state, new DealCommand.Take(1)));
        final DealState afterFirstPass = applied(engine.apply(announced, new DealCommand.Pass(0)));
        final DealState next = applied(engine.apply(afterFirstPass, new DealCommand.Pass(2)));

        assertThat(next.roundStarterSeat()).isEqualTo(2);
        assertThat(next.defenderSeat()).isZero();
    }

    @DisplayName("Should refuse any command When the deal is already over")
    @Test
    void shouldRefuseAnyCommandWhenTheDealIsAlreadyOver() {
        final DealState state = aDeal().withPhase(DealPhase.DEAL_OVER).withHand(0, ACE_CLUBS).build();

        assertThat(engine.apply(state, new DealCommand.Attack(0, ACE_CLUBS)))
                .isEqualTo(MoveResult.rejected(RejectionReason.NOT_YOUR_TURN));
    }

    @DisplayName("Should reveal the face-down card When it is played as the last card")
    @Test
    void shouldRevealTheFaceDownCardWhenItIsPlayedAsTheLastCard() {
        final DealState state = aDeal()
                .withEmptyDeck()
                .withFaceDownCard(0, SEVEN_DIAMONDS)
                .withHand(1, ACE_CLUBS)
                .build();

        final MoveResult result = engine.apply(state, new DealCommand.RevealFaceDown(0));

        final DealState next = applied(result);
        assertThat(next.playerAt(0).hasFaceDownCard()).isFalse();
        assertThat(next.table()).singleElement()
                .satisfies(slot -> assertThat(slot.attack()).isEqualTo(SEVEN_DIAMONDS));
        assertThat(events(result)).containsExactly(
                new DealEvent.FaceDownRevealed(0, SEVEN_DIAMONDS),
                new DealEvent.CardAttacked(0, SEVEN_DIAMONDS));
    }

    private static DealState applied(final MoveResult result) {
        assertThat(result).isInstanceOf(MoveResult.Applied.class);
        return ((MoveResult.Applied) result).state();
    }

    private static List<DealEvent> events(final MoveResult result) {
        return ((MoveResult.Applied) result).events();
    }
}

File: back-go/internal/domain/game/deal_engine_test.go
package game

import "testing"

// Перенос DealEngineTest и DealRefillAndExitTest — проверки самого автомата раздачи.
//
// ⭐ Здесь живёт починка тупика (ADR-051): состояние, из которого ни один игрок не мог
// сходить, а раздача не была окончена. Тесты в конце файла держат именно её.

func engine() DealEngine { return NewDefaultDealEngine() }

func mustApply(t *testing.T, state DealState, cmd DealCommand) MoveResult {
	t.Helper()
	result := engine().Apply(state, cmd)
	if !result.Applied {
		t.Fatalf("команда отклонена с причиной %s, а должна была примениться", result.Reason)
	}
	return result
}

func mustReject(t *testing.T, state DealState, cmd DealCommand, reason RejectionReason) MoveResult {
	t.Helper()
	result := engine().Apply(state, cmd)
	if result.Applied {
		t.Fatalf("команда применена, а должна была быть отклонена (%s)", reason)
	}
	if result.Reason != reason {
		t.Fatalf("причина отказа %s, ждали %s", result.Reason, reason)
	}
	return result
}

func TestAttackMovesCardFromHandToTable(t *testing.T) {
	six := NewPip(Six, Clubs)
	state := aDeal().withHand(0, six).withHand(1, NewPip(Ace, Clubs)).build()

	result := mustApply(t, state, AttackCommand{Seat: 0, Card: six})

	if len(result.State.Table) != 1 || result.State.Table[0].Attack != six {
		t.Error("карта не легла на стол")
	}
	if result.State.MustPlayerAt(0).HoldsInHand(six) {
		t.Error("карта осталась в руке — она должна была уйти на стол")
	}
	if result.State.Phase != PhaseDefend {
		t.Errorf("фаза %s, ждали DEFEND", result.State.Phase)
	}
}

// ⚠️ Отклонённая команда не меняет состояние ВООБЩЕ. Иначе отказ оставлял бы следы,
// и партия расходилась бы с тем, что видели игроки.
func TestRejectedCommandLeavesStateUntouched(t *testing.T) {
	six := NewPip(Six, Clubs)
	state := aDeal().withHand(0, six).withHand(1, NewPip(Ace, Clubs)).build()

	result := engine().Apply(state, AttackCommand{Seat: 2, Card: six})

	if result.Applied {
		t.Fatal("ход не с того места должен отклоняться")
	}
	if len(result.State.Table) != 0 || len(result.State.Players) != 0 {
		t.Error("отклонённая команда вернула состояние — она не должна возвращать ничего")
	}
	if len(state.Table) != 0 {
		t.Error("исходное состояние изменилось")
	}
}

func TestDefenceMarksNamedSlotBeaten(t *testing.T) {
	target := NewPip(Seven, Diamonds)
	card := NewPip(Nine, Diamonds)
	state := aDeal().withHand(1, card).withAttack(target).withPhase(PhaseDefend).build()

	result := mustApply(t, state, DefendCommand{Seat: 1, Card: card, Target: target})

	if !result.State.Table[0].IsBeaten() {
		t.Error("слот не отмечен отбитым")
	}
	if result.State.Table[0].Defence != card {
		t.Error("отбито не той картой, что назвали")
	}
}

func TestStaysInDefenceWhileUnbeatenRemain(t *testing.T) {
	first := NewPip(Seven, Diamonds)
	second := NewPip(Seven, Clubs)
	card := NewPip(Nine, Diamonds)
	state := aDeal().withHand(1, card, NewPip(Ace, Hearts)).
		withAttack(first, second).withPhase(PhaseDefend).build()

	result := mustApply(t, state, DefendCommand{Seat: 1, Card: card, Target: first})

	if result.State.Phase != PhaseDefend {
		t.Errorf("фаза %s, а на столе осталась неотбитая карта", result.State.Phase)
	}
}

// Перевод делает переводящего атакующим, а защиту сдвигает по кругу.
func TestTransferMakesTransferrerTheAttacker(t *testing.T) {
	card := NewPip(Seven, Clubs)
	state := aDeal().withPlayers(3).
		withHand(1, card).
		withHand(2, NewPip(Ace, Hearts), NewPip(King, Hearts), NewPip(Queen, Hearts)).
		withAttack(NewPip(Seven, Diamonds)).withPhase(PhaseDefend).build()

	result := mustApply(t, state, TransferCommand{Seat: 1, Card: card})

	if result.State.AttackRightSeat != 1 || result.State.RoundStarterSeat != 1 {
		t.Error("переводящий обязан стать атакующим")
	}
	if result.State.DefenderSeat != 2 {
		t.Errorf("защита ушла на место %d, ждали 2", result.State.DefenderSeat)
	}
	if len(result.State.PassedSeats) != 0 {
		t.Error("перевод открывает раунд заново — пасы обнуляются")
	}
}

// ⭐ Право подкидывать уходит ВТОРОМУ СОСЕДУ, то есть следующему за защищающимся.
func TestPassHandsRightToSecondNeighbour(t *testing.T) {
	state := aDeal().withPlayers(3).
		withHand(0, NewPip(Six, Clubs)).
		withHand(2, NewPip(Six, Hearts)).
		withAttack(NewPip(Seven, Diamonds)).withPhase(PhaseDefend).build()

	result := mustApply(t, state, PassCommand{Seat: 0})

	if result.State.AttackRightSeat != 2 {
		t.Errorf("право у места %d, ждали второго соседа — место 2", result.State.AttackRightSeat)
	}
}

// ⚠️ Право назад НЕ возвращается: спасовавший в этом раунде больше не подкидывает.
func TestPassedPlayerNeverGetsTheRightBack(t *testing.T) {
	six := NewPip(Six, Clubs)
	state := aDeal().withPlayers(3).
		withHand(0, six).
		withAttack(NewPip(Seven, Diamonds)).
		withPassed(0).withAttackRight(0).withPhase(PhaseDefend).build()

	mustReject(t, state, AttackCommand{Seat: 0, Card: six}, NotYourTurn)
}

// «Беру» раунд не закрывает: подкидывающие докидывают, пока не спасуют.
func TestTakeKeepsRoundAliveForFollowUps(t *testing.T) {
	state := aDeal().withPlayers(3).
		withHand(0, NewPip(Seven, Clubs)).
		withHand(1, NewPip(Ace, Hearts)).
		withAttack(NewPip(Seven, Diamonds)).withPhase(PhaseDefend).build()

	result := mustApply(t, state, TakeCommand{Seat: 1})

	if result.State.Phase != PhaseTaking {
		t.Errorf("фаза %s, ждали TAKING", result.State.Phase)
	}
	if len(result.State.Table) == 0 {
		t.Error("стол уехал сразу — а подкидывающие ещё не спасовали")
	}
}

// ⭐ После «беру» рука взявшего больше не ограничивает подкид: он всё равно заберёт всё.
func TestTakersHandSizeIsIgnoredForFollowUps(t *testing.T) {
	extra := NewPip(Seven, Clubs)
	state := aDeal().withPlayers(3).
		withHand(0, extra).
		withHand(1, NewPip(Ace, Hearts)). // одна карта
		withAttack(NewPip(Seven, Diamonds)).
		withPhase(PhaseTaking).withAttackRight(0).build()

	mustApply(t, state, AttackCommand{Seat: 0, Card: extra})
}

// ...но потолок раунда остаётся.
func TestFollowUpsStillCappedByRoundLimit(t *testing.T) {
	extra := NewPip(Six, Clubs)
	onTable := []Card{
		NewPip(Six, Diamonds), NewPip(Six, Hearts), NewPip(Six, Spades),
		NewPip(Seven, Diamonds), NewPip(Seven, Hearts),
	}
	state := aDeal().withPlayers(3).
		withHand(0, extra).
		withHand(1, NewPip(Ace, Hearts)).
		withAttack(onTable...).
		withPhase(PhaseTaking).withAttackRight(0).build()

	mustReject(t, state, AttackCommand{Seat: 0, Card: extra}, AttackLimitReached)
}

func TestDefenceAndTransferRefusedAfterTake(t *testing.T) {
	target := NewPip(Seven, Diamonds)
	card := NewPip(Nine, Diamonds)
	state := aDeal().withPlayers(3).withHand(1, card).
		withAttack(target).withPhase(PhaseTaking).build()

	mustReject(t, state, DefendCommand{Seat: 1, Card: card, Target: target}, DefenderAlreadyTook)
	mustReject(t, state, TransferCommand{Seat: 1, Card: NewPip(Seven, Clubs)}, DefenderAlreadyTook)
}

func TestNothingToTakeOnEmptyTable(t *testing.T) {
	state := aDeal().withPhase(PhaseAttack).build()

	mustReject(t, state, TakeCommand{Seat: 1}, NothingToTake)
}

func TestEveryCommandRefusedWhenDealIsOver(t *testing.T) {
	six := NewPip(Six, Clubs)
	state := aDeal().withHand(0, six).withPhase(PhaseDealOver).build()

	mustReject(t, state, AttackCommand{Seat: 0, Card: six}, NotYourTurn)
	mustReject(t, state, PassCommand{Seat: 0}, NotYourTurn)
	mustReject(t, state, TakeCommand{Seat: 1}, NotYourTurn)
}

// ── Тупик, ради которого всё и чинилось (ADR-051) ────────────────────────────

// ⚠️ Отбита последняя карта, а право подкидывать осталось за СПАСОВАВШИМ. Он не может
// ни подкинуть, ни спасовать снова — и раздача вставала намертво У ВСЕХ за столом.
//
// Проверяем не «у кого-то есть ход», а именно то, что ход есть у обладателя права:
// первая версия этого теста проходила на сломанном коде, потому что у защиты
// оставалось бессмысленное «беру».
func TestNoDeadlockAfterLastCardBeaten(t *testing.T) {
	attack := NewPip(Seven, Diamonds)
	defence := NewPip(Nine, Diamonds)

	state := aDeal().withPlayers(3).
		withHand(0, NewPip(King, Clubs)).
		withHand(1, defence).
		withHand(2, NewPip(Queen, Clubs)).
		withAttack(attack).
		withPassed(0).withAttackRight(0).
		withPhase(PhaseDefend).
		build()

	result := mustApply(t, state, DefendCommand{Seat: 1, Card: defence, Target: attack})
	next := result.State

	if next.Phase == PhaseDealOver {
		return // раздача закрылась — тупика нет по определению
	}

	// ⭐ Главное утверждение: у того, за кем право, есть хотя бы один законный ход.
	holder := next.AttackRightSeat
	if next.HasPassed(holder) {
		t.Fatalf("право осталось за спасовавшим местом %d — это и есть тупик", holder)
	}

	rules := NewMoveRules(DefaultRulesConfig())
	hasMove := false
	for _, card := range next.MustPlayerAt(holder).Hand {
		if rules.CanAttack(next, holder, card).IsAllowed() {
			hasMove = true
			break
		}
	}
	if !hasMove && engine().Apply(next, PassCommand{Seat: holder}).Applied {
		hasMove = true
	}
	if !hasMove {
		t.Fatalf("у обладателя права (место %d) нет ни одного законного хода — стол встал", holder)
	}
}

// Раунд закрывается, когда отбита последняя карта и подкидывать больше некому.
func TestRoundClosesWhenNobodyCanFollowUp(t *testing.T) {
	attack := NewPip(Seven, Diamonds)
	defence := NewPip(Nine, Diamonds)

	state := aDeal().withPlayers(2).
		withHand(0, NewPip(King, Clubs)).
		withHand(1, defence).
		withAttack(attack).
		withPassed(0).withAttackRight(0).withDefender(1).
		withPhase(PhaseDefend).
		build()

	result := mustApply(t, state, DefendCommand{Seat: 1, Card: defence, Target: attack})

	if len(result.State.Table) != 0 {
		t.Error("стол не уехал в отбой, хотя подкидывать некому")
	}
	if !hasEvent(result.Events, func(e DealEvent) bool { _, ok := e.(RoundBeaten); return ok }) {
		t.Error("нет события «бито»")
	}
}

// ── Добор и выход ────────────────────────────────────────────────────────────

// ⚠️ Порядок добора не формальность: колода конечна, и начавший раунд берёт первым.
func TestScarceCardsGoToRoundStarterFirst(t *testing.T) {
	attack := NewPip(Seven, Diamonds)
	defence := NewPip(Nine, Diamonds)

	state := aDeal().withPlayers(2).
		withHand(0, NewPip(King, Clubs)).
		withHand(1, defence).
		withAttack(attack).
		withDeck(NewPip(Ace, Clubs)). // ровно одна карта на двоих
		withPassed(0).withAttackRight(0).withDefender(1).
		withRoundStarter(0).
		withPhase(PhaseDefend).
		build()

	result := mustApply(t, state, DefendCommand{Seat: 1, Card: defence, Target: attack})

	// Место 0 начало раунд — единственная карта его.
	if result.State.MustPlayerAt(0).HandSize() != 2 {
		t.Errorf("у начавшего раунд %d карт, он должен был добрать первым",
			result.State.MustPlayerAt(0).HandSize())
	}
}

func hasEvent(events []DealEvent, match func(DealEvent) bool) bool {
	for _, event := range events {
		if match(event) {
			return true
		}
	}
	return false
}
